### PR TITLE
Profiling new

### PR DESCRIPTION
**Pull request by neworderofjamie**

### CODE REVIEW

File: nengo_spinnaker/operators/lif.py [hunk 1]
```diff
@@ -234,17 +265,31 @@ def after_simulation(self, netlist, simulator, n_steps):
...
+        # If profiling is enabled
+        if self.profiler_region.n_samples > 0:
+            # Loop through vertices
+            for vertex in self.vertices:
+                # Get profiler output memory block
+                mem = self.profiler_output_mem[vertex]
+                mem.seek(0)
+
+                # Read profiler data from memory and put somewhere accessible
+                simulator.profiler_data[self.ensemble] =\
```
Comment (mundya) [inline nengo_spinnaker/operators/lif.py:282]: Worth noting that for `sim.run(t)` if `t > sim.period` this line will be called multiple times and will overwrite stored values.  Don't know if that's a concern provided we know how long our experiments will be.

File: nengo_spinnaker/regions/profiler.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    def write_subregion_to_file(self, *args, **kwargs):  # pragma: no cover
+        pass  # Nothing to do
+
+    def read_from_mem(self, mem, tag_names):
+        # Read number of words written by profiler
+        word_written = struct.unpack("I", mem.read(4))[0]
+
+        # Read these from memory
+        data = np.fromstring(mem.read(word_written * 4), dtype=np.uint64)
```
Comment (mundya) [inline nengo_spinnaker/regions/profiler.py]: Why is the `dtype=uint64`?

File: nengo_spinnaker/utils/profiling.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        total_sample_duration_per_timestep = np.zeros(last_sample_time)
+        for sample_duration, index in zip(times[1], sample_timestep_indices):
+            total_sample_duration_per_timestep[index] += sample_duration
+
+        print("\tMean time per timestep:%fms" %
+              (np.average(total_sample_duration_per_timestep)))
+
+
+def write_csv_header(profiling_data, csv_writer, extra_column_headers):
+    # Write header row with extra column headers
```
Comment (mundya) [inline nengo_spinnaker/utils/profiling.py]: Should be a docstring?

File: nengo_spinnaker/utils/profiling.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    # followed by tag names found in profiling_data
+    csv_writer.writerow(extra_column_headers + list(profiling_data.iterkeys()))
+
+
+def write_csv_row(profiling_data, csv_writer, extra_column_values):
```
Comment (mundya) [inline nengo_spinnaker/utils/profiling.py:48]: Should have a docstring?

File: nengo_spinnaker/utils/profiling.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import numpy as np
+
+
+def print_summary(profiling_data, duration):
```
Comment (mundya) [inline nengo_spinnaker/utils/profiling.py]: Should have a docstring?

File: spinnaker_components/Makefile.common [hunk 6]
```diff
@@ -163,7 +172,7 @@ endif
 	$(LS) $(APP_DIR)/$(APP).aplx
 
 
-# Build the ELF file
+# Build the ELF filBUILDDIRe
```
Comment (mundya) [inline spinnaker_components/Makefile.common]: Yay search & replace

**Comment by mundya** [comment]
flake8 :heart: you

File: nengo_spinnaker/regions/profiler.py (hunk 2 above)
Comment (neworderofjamie) [inline nengo_spinnaker/regions/profiler.py]: Yeah, fine point - this could be read in uint32s to start with

Comment (mundya) [inline nengo_spinnaker/regions/profiler.py]: I'm guessing that `word_written` is always even?

Comment (neworderofjamie) [inline nengo_spinnaker/regions/profiler.py]: Yes it is - the profiler writes out pairs of uint32s

**Comment by mundya** [comment]
Thanks @neworderofjamie

**Pull request merged**